### ISSUE TITLE
Problem with RL-tuner code.

### ISSUE THREAD

**Issue by hany-tawfik** [tensorflow/magenta#1116]
I am having a problem running the RL-tuner Jupyter notebook. At first, it is not opening at all, but then I have found that there is a comma on line 13 that should be deleted. Afterwards, when I start running the code, it downloads the note_rnn.ckpt file and then I get this error message: NotFoundError (see above for traceback): Tensor name "rnn_model/rnn/multi_rnn_cell/cell_0/lstm_cell/bias" not found in checkpoint files /Users/HanyTawfik/note_rnn.ckpt. I also get the same error when I run “rl_tuner_train” from the terminal.

I then switched to using basic_rnn instead -as mentioned in the documentation- however, I got this error: TypeError: Error converting shape to a TensorShape: int() argument must be a string or a number, not 'tuple'.

I very much like this model and I want to use it in my master thesis, please advice.

**Comment by ml4046** [comment]
The main issue here is that `tf.placeholder` is reading in a tuple. The code is trying to set up the initial state such that `MultiRNNCell` has dynamic input size. Usually it is initialized with `zero_state` but requires you to specify a batch_size. You can alternatively create a placeholder to specify the batch_size and modify the code such that you are also passing in the batch_size during a session.

**Comment by miraodasilva** [comment]
I didn't really understand the exact solution to this @ml4046 . Did you end up solving the issue @hany-tawfik ? I'm having the same exact problem.

**Comment by miraodasilva** [comment]
Thank you very much for the response. Do you have any idea why this error exists in the first place? Is it a matter of the code being somewhat old?

**Comment by miraodasilva** [comment]
@ml4046 To be clear, I am under the impression that the problem is that cell.state_size used to return an integer and now returns a tuple, and that's why the code does not work. In any case, the function to get_zero_state also does not work for the same reason (it returns "np.zeros((self.batch_size, self.cell.state_size))"). In other words, it again seems to assume that the state_size is an integer. 

I apologize if I am completely misunderstanding, I am very new to Tensorflow. 

Thanks

**Comment by miraodasilva** [comment]
Okay, I think that worked! Did you also have the same trouble with the checkpoint file for the basic RNN after this?

**Comment by ml4046** [comment]
@miraodasilva I did so I just trained my own model with `melody_rnn` and tuned with `rl_tuner` which worked out for me (remember to pass custom hparams for the melody rnn when you initialize `rl_tuner`)

**Comment by ml4046** [comment]
@miraodasilva No worries, from what I remember, you have to pass in a custom `note_rnn_hparams` when you initialize `rl_tuner` is because the initialization misses the flag that you're using a custom trained model so it will initialize from the default even if you specify a custom checkpoint directory.

As with using the `rl_tuner` for tuning the default `basic_rnn`, I haven't tested out the pre-config model since I only have been tuning my own 'basic_rnn' generated from `melody_rnn`

**Comment by miraodasilva** [comment]
@ml4046 
I converted the checkpoint to the right format successfully. However, now I get this error. Did you have this same error by any chance? Thank you very much. 

"TypeError: The value of a feed cannot be a tf.Tensor object. Acceptable feed values include Python scalars, strings, lists, numpy ndarrays, or TensorHandles.For reference, the tensor object was Tensor("MultiRNNCellZeroState_2/DropoutWrapperZeroState_1/BasicLSTMCellZeroState/zeros:0", shape=(128, 512), dtype=float32) which was passed to the feed with key Tensor("q_network/MultiRNNCellZeroState/DropoutWrapperZeroState_1/BasicLSTMCellZeroState/zeros:0", shape=(128, 512), dtype=float32)."

**Comment by ml4046** [comment]
@miraodasilva 
It's hard to tell which line is causing this but the general idea is you're feeding in a tensor when you run your sessions from `session.run(...,feed_dict:{state:<some tf.Tensor>}).` You want to feed in the values instead (i.e. samples from the replays), not the Tensor itself.

**Comment by miraodasilva** [comment]
@ml4046  The issue is when I am passing the state_value as the initial state in the feed_dict. The state of the LSTM is a tuple of LSTMSTateTuples of Tensors, and therefore I get the error. The exact line of the error is 1789 I believe, where this happens:

softmax, self.q_network.state_value = self.session.run(
            [self.action_softmax, self.q_network.state_tensor],
            {self.q_network.melody_sequence: input_batch,
             self.q_network.initial_state: self.q_network.state_value, << PROBLEM
             self.q_network.lengths: lengths})

I don't understand how to solve it. If I try to use eval on the tensors it tells me that the session I am using is not the same as the Tensor's session. From my understanding, if I could  transform the state_value into a raw format instead of Tensor, the problem would be solved, but I can't seem to do this successfully.

**Comment by miraodasilva** [comment]
This also happens during training in line 704, where again the state_value is fed into the feed_dict and the same error appears.

**Comment by ml4046** [comment]
@miraodasilva Yes, try creating a placeholder for specifying the and passing it in to your sessions. You'll need this when you sample from the replay buffer since when you first initialize your state, the batchsize is 1

**Comment by miraodasilva** [comment]
@ml4046 I got it to work! I am currently using the Basic RNN by the way, not the NoteRNN. It is generating proper MIDI outputs. By the way, I am actually editing the file in anaconda/envs/magenta...../rl_tuner.py and then running "python rl_tuner.py". Is there a better way to do this? I apologize if this is a pretty ridiculous question.

Thank you very much for the help.

**Comment by miraodasilva** [comment]
@ml4046 It turns out that I was generating but without training (tuning) with RL at all. When I run the train function it gives me some errors which again have to do with the new form of LSTM states. Did you have to modify the training_step/train functions to accomodate the new form of LSTM States? I am removing all of the .flatten from the states and chaging the way they are concatenated when they are fed in batches during the training step. It is a bit of work but I suppose it is necessary to make it work. Let me know if you have a different solution for this.

Thank you for your time.

**Comment by ml4046** [comment]
@miraodasilva 

Yes I modified how values sampled from the replay buffer are fed into the session like you mentioned.

**Comment by miraodasilva** [comment]
@ml4046 It's working and training successfully after some modifcations! The samples post-training actually sound quite good, but they are lacking pauses or continued notes (0s and 1s in their notation). This can be easily observed in the note probabilities pre and post RL. Did this happen to you as well by any chance?

Again, thank you so much for the help!

**Comment by ml4046** [comment]
@miraodasilva 
I had a good amount of pauses and no actions actually and I didn't modify the reward function. Maybe you can try modifying the music reward function? Or maybe it has to do with the note sequences you generated for training previously?

No problem at all

**Comment by miraodasilva** [comment]
@ml4046 
Hi! I have been busy with other stuff such as the background report for my thesis, but I managed to get the whole thing working, train a basic_rnn and use my own dataset. However, I am now trying to adapt the code to use an attention_rnn instead (a variant of the basic_rnn). I have two issues: 1. I get Nan loss during training for the attention_rnn and 2. The checkpoint that I have (from earlier in the training, before the error) does not work (shapes are different). What I did was simply change the make_rnn_cell part of the code to include attn_length=40, in order to apply the attention wrapper on the lstm cell. 

Anyway, have you experienced any issues similar to these with Magenta code? 

As always, thanks so much for the help.

**Comment by 4Akera** [comment]
Creating a placeholder doesn't simply remove the problem.

**Comment by miraodasilva** [comment]
Hi,
Really sorry about this but I have not actively worked on this code for a while and I have now moved on to working almost entirely on Pytorch projects, so I'm really not familiar with this issue anymore. I can't remember if I ended up having a solution for this, or what the solution was. I would suggesting trying to contact the magenta devs or contacting the user who made a pull request about his recently. I hope you can find a solution, good luck.

**Comment by 4Akera** [comment]
No problem, and yeah it's is an old stuff. However, I will try to investigate the issue carefully and ask for help as you suggest. 
Thanks for your response.  
Wish you good luck too!

**Comment by Vebrun** [comment]
> Okay, I think that worked! Did you also have the same trouble with the checkpoint file for the basic RNN after this?



> Hello, how exactly do I need to modify the code for the placeholder problem? From your conversations, I do not understand the specific solution
